### ISSUE TITLE
Documentation built on CI should be published immediately

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#1702]
It would be useful to have the documentation of the dev version always available online. We already build the documentation on CI. This issue is to push the built documentation to a server immediately, as it gets built.

**Comment by szhorvat** [comment]
@iosonofabio You mentioned in #1276 that you could take a look at this. Would you have time to do so?

**Comment by vtraag** [comment]
I think it should be fairly straightforward to setup using GitHub actions. I also have setup something similar for the R interface (pushing build source package towards the `master` branch), see https://github.com/igraph/rigraph/blob/dev/.github/workflows/build-and-check.yml.

**Comment by iosonofabio** [comment]
No problem, I'll take a look

**Comment by iosonofabio** [comment]
Here an example:

https://tigeroakes.com/posts/til-github-pages-two-branches-github-actions/

The way that would work is by building both master and develop branch at once and getting Github Pages to do the lifting onto the internet. In other words, we could make a single Github Action that is triggered on both branches and refreshes the whole website. The develop docs would go under a subfolder, e.g. igraph.org/develop/...

Alternatively, we could try and use subdomains, i.e. develop.igraph.org/... then we would have to link the "develop" subdomain with a (currently inexistant) develop branch on the igraph/igraph.org repo, and the normal subdomains (i.e. the main domain, the "www" subdomain) to the "master" branch on igraph/igraph.org.

@ntamas any preferences here? I'm asking because you're the one with privileges to alter the CNAME DNS records and similar.

**Comment by ntamas** [comment]
I don't have access to the CNAME DNS records, only @gaborcsardi has. A subdomain does not really make sense to me, though, because the homepage hosts the documentation of the C core, the Python interface and the R interface and there isn't any consistent concept of a `develop` branch that would apply to all three repos.

Also, @szhorvat has been asking me for long to somehow make it possible to browse older versions of the documentation as well. This would also influence the URL structure so we should have something like `igraph.org/c/doc/VERSION/` for a given version of the C core, and then we could use a `stable` symlink to the latest stable version, and `develop` could point to the documentation built from the head of the development branch. (But then what about the head of the `master` branch, which could also be ahead of `stable`?). The same pattern could apply for `igraph.org/r/doc/VERSION/` and `igraph.org/python/doc/VERSION/`.

At this stage I was wondering whether this structure starts to become too complex to be managed by Github Pages (or, if we would hit some kind of a repo size limit with the `gh-pages` branch). Maybe we should just rent space from Amazon S3 and rsync the built documentation there, separated by version numbers? The idea would be that Jekyll would keep on generating the main website with Github Pages, but we would have a separate `doc.igraph.org` domain (backed by S3), and Github Actions would update that as needed?

**Comment by iosonofabio** [comment]
Ok then more option 1 than 2.

The igraph/igraph.org repo already has a `.nojekyll`, so it's not actually doing much except providing the files ready to be served. So the architecture would be for each repo (e.g. `python-igraph`, `rigraph`) to pull the latest commit from `igraph/igraph.org`, add the docs from this build of that particular repo without touching anything else, and then push back to `igraph/igraph.org`. If CI for two different repos/branches are happening in parallel, there is a race condition though.

If we were to just have an S3 bucket and each repo (e.g. `rigraph`) dumps a bunch of files there, there is less of an issue with parallel dumping, as long as they don't touch each other's folders (they shouldn't). But a bucket is not versioned as well as a github repo, so if anything happens to it we might have a hard time recovering the history.

One way to keep using Github Pages for now while dealing with the race condition would be a PR. So each repo could open a PR on `igraph/igraph.org` upon documentation completion, and that would trigger an auromatic merge by the `igraph/igraph.org` repo.

I'll try play with these options in the next few days and report back.

**Comment by gaborcsardi** [comment]
> If CI for two different repos/branches are happening in parallel, there is a race condition though.

You can use this to avoid the race condition: https://github.com/softprops/turnstyle I would fork it into out org and use the fork, just to be safe.

**Comment by iosonofabio** [comment]
thank you, will look into it!

**Comment by szhorvat** [comment]
Note: don't use Travis to do this (not that it's likely we would have).

This was shared by Tamás today: https://arstechnica.com/information-technology/2021/09/travis-ci-flaw-exposed-secrets-for-thousands-of-open-source-projects/

**Comment by szhorvat** [comment]
> No problem, I'll take a look

It'd be great to have this set up this autumn, if you have the time, Fabio.

**Comment by iosonofabio** [comment]
Thanks for reminding, not exactly on top of my to-do list ATM, and it's quite a bunch of moving parts. But yes it'd be nice to have at some point. Want to give it a go yourself?

**Comment by szhorvat** [comment]
@iosonofabio Since you're working with docs recently, if you feel you have some spare capacity and the task is not too difficult, this would be quite nice to have. Not a top priority of course, but it's helpful to have the latest docs pre-generated and already when making small fixes.

**Comment by ntamas** [comment]
Hmmmm, thinking out loud: this would mean that in our own CI build, we also build the documentation, bundle it up nicely as HTML, then copy it over somehow to the `igraph/igraph.org` repo and replace the current "latest" docs with that, and run Jekyll on it.

At this stage I think we are hittng the limits of the current Jekyll / Github Pages setup and we should think about moving on to something else. For instance, building the docs and then pushing them to an Amazon S3 bucket is probably way easier (if we can let go of the homepage-specific styling), and then we could link to that bucket from the homepage instead. How does that sound?

**Comment by szhorvat** [comment]
It would serve the purpose for me.

**Comment by iosonofabio** [comment]
Thanks for the input. Doing this properly outside of a set infra (e.g. RTD) is tricky because the new docs must include html links to all older versions: i.e. they need to be aware of that context at build time on a CI runner. There are ways to do it (e.g. scraping our own official docs, checking git tag history) but it's unclear *we* are the ones who should do it.

The alternative I'd prefer would be to migrate the C docs onto a platform that supports versioned docs natively (e.g. RTD but for C) and has a we hook into our C core repo so everything is published upon changes (e.g. push, PR, tagging, etc.).

One *might* be able to use RTD itself for this, they have experimental support for custom build commands (as opposed to sphinx). However, I'm planning to shop around for more commonly used platforms first. If you have any suggestions, I'm all ears 😊

**Comment by szhorvat** [comment]
To be clear, I am looking for a tool to make development easier. I was not looking for a user-facing, nicely formatted setup. I just want to reduce the number of times I have to build docs locally. The "develop" section on igraph.org is already very helpful, but having more frequent updated would be nicer. If these frequent updates are pushed to a different location, without fancy formatting, and without cross-linking, that is perfectly fine for me.
